core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala

```diff
@@ -128,17 +161,31 @@ class ShuffleMemoryManager protected (
       metrics.updatedBlocks = Some(lastUpdatedBlocks ++ evictedBlocks.toSeq)
     }
     taskMemory(taskAttemptId) += acquired
+    val consumers = taskConsumers.getOrElseUpdate(taskAttemptId, new mutable.HashMap())
+    consumers += consumer -> (consumers.getOrElse(consumer, 0L) + acquired)
     acquired
   }
 
+  def release(numBytes: Long): Unit = release(numBytes, null)
+
   /** Release numBytes bytes for the current task. */
-  def release(numBytes: Long): Unit = memoryManager.synchronized {
+  def release(numBytes: Long, consumer: MemoryConsumer): Unit = memoryManager.synchronized {
     val taskAttemptId = currentTaskAttemptId()
     val curMem = taskMemory.getOrElse(taskAttemptId, 0L)
     if (curMem < numBytes) {
       throw new SparkException(
         s"Internal error: release called on $numBytes bytes but task only has $curMem")
     }
+    taskConsumers.get(taskAttemptId).map { consumers =>
+      if (consumers.contains(consumer)) {
+        consumers(consumer) -= numBytes
+        if (consumers(consumer) <= 0) {
+          consumers.remove(consumer)
+        }
+      } else {
+        logError(s"BUG: release $numBytes from $consumer in task $taskAttemptId")
```

Comment:
If the `spark.testing` system property is set, then I think this warning should become an exception, since that will let us detect problems in unit tests.